cpp/oneapi/dal/backend/primitives/distance/correlation_distance_misc_dpc.cpp

```diff
@@ -0,0 +1,95 @@
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/detail/profiler.hpp"
+
+#include "oneapi/dal/backend/primitives/distance/correlation_distance_misc.hpp"
+
+#include "oneapi/dal/backend/primitives/blas.hpp"
+#include "oneapi/dal/backend/primitives/reduction.hpp"
+#include "oneapi/dal/backend/primitives/stat/cov.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+template <typename Float, ndorder order>
+sycl::event compute_deviation(sycl::queue& q,
+                              const ndview<Float, 2, order>& inp,
+                              ndview<Float, 2>& out,
+                              const event_vector& deps) {
+    ONEDAL_ASSERT(inp.has_data());
+    ONEDAL_ASSERT(out.has_mutable_data());
+    const auto n = out.get_dimension(0);
+    const auto p = out.get_dimension(1);
+    ONEDAL_ASSERT(n == inp.get_dimension(0));
+    ONEDAL_ASSERT(p == inp.get_dimension(1));
+    auto* const inp_ptr = inp.get_mutable_data();
+    const auto out_stride = out.get_leading_stride();
+    auto* const out_ptr = out.get_mutable_data();
+    auto out_range = make_range_2d(n, p);
+    auto inp_sum = ndarray<Float, 1>::empty(q, { n });
+    auto inp_mean = ndarray<Float, 1>::empty(q, { n });
+
+    // Collect sums of each row of input matrix
+    auto sum_event = reduce_by_rows(q, inp, inp_sum, sum<Float>{}, identity<Float>{}, deps);
+
+    // Compute mean of each row of input matrix using sum event
+    auto mean_event = means(q, p, inp_sum, inp_mean, { sum_event });
+
+    // Return event that updates output matrix with centered values (input(x) - input_mean(x))
+    return q.submit([&](sycl::handler& h) {
+        h.depends_on({ mean_event });
+        auto const inp_mean_acc = inp_mean.get_data();
+        h.parallel_for(out_range, [=](sycl::id<2> idx) {
+            const auto offset = idx[0] * out_stride + idx[1];
+            out_ptr[offset] = inp_ptr[offset] - inp_mean_acc[idx[0]];
+        });
+    });
+}
+
+template <typename Float, ndorder order>
+std::tuple<ndarray<Float, 2>, sycl::event> compute_deviation(sycl::queue& q,
+                                                             const ndview<Float, 2, order>& inp,
+                                                             const event_vector& deps,
+                                                             const sycl::usm::alloc& alloc) {
+    const auto n = inp.get_dimension(0);
+    const auto p = inp.get_dimension(1);
+    auto res_array = ndarray<Float, 2>::empty(q, { n, p }, alloc);
```

Comment:
not sure, but looks like a potential bug, due to ndarray<Float, 2>::empty(q, { n, p }, alloc::host); can potentially calls an issue. Can you please doublecheck such case? May be applicable for other distances

Reply:
https://github.com/uxlfoundation/oneDAL/blob/7cc1c9b090d8291abad40063b0a713a1b516de18/cpp/oneapi/dal/backend/primitives/distance/cosine_distance_misc_dpc.cpp#L57-L65
What is the bug exactly? I also modeled this after the `compute_inversed_l2_norms()` function, except I converted it into a 2d array.

Reply:
I was thinking that potentially 
[[maybe_unused]] auto res_array_check = ndarray<Float, 1>::empty(q, { n_samples }, sycl::usm::alloc::host); - incorrect api
[[maybe_unused]] auto res_array_check = ndarray<Float, 1>::empty({ n_samples }, sycl::usm::alloc::host); - correct
[[maybe_unused]] auto res_array_check = ndarray<Float, 1>::empty(q, { n_samples }, sycl::usm::alloc::shared/device); - correct

but i doublechecked and it works